### ISSUE TITLE
Using in a multi-tenant Application

### ISSUE THREAD

**Issue by OPVL** [rootinc/laravel-azure-middleware#18]
Hi,

I was hoping to use this in an internal application that we would allow our clients access to. 

I'm attempting to access the **Graph API** to persist basic user data and then build out our own roles/permissions logic to avoid having to use the Enterprise solution through azure.

To get this working so far I've set `AZURE_TENANT_ID=common` and `AZURE_RESOURCE=https://graph.windows.net` however I'm running in to a 401 using the [extended installation](https://github.com/rootinc/laravel-azure-middleware/#extended-installation) with the **Invalid Audience** message returned when trying to get a users profile after successfully logging in. When setting the audience to graph.microsoft.com I get a generic audience not found under tenant error however the .net seems to work. Not sure if this is relevant but thought it would help to include it.

Is there something that I'm additionally in need of in order to make this work in a multi tenanted application or am I being ultra dense and failing to grasp something painfully obvious.

tldr: I've set tenant ID to common for authentication (which works great) however when I try and get a user profile after authenticating I run in a 401 (**Invalid Audience**)

Below are screenshots of the (truncated) error returned and my decoded JWT token.

![image](https://user-images.githubusercontent.com/17033678/68291943-cd663f00-0082-11ea-8951-99a4e3942b77.png)

```json
{
  "typ": "JWT",
  "alg": "RS256",
  "x5t": "aPctw_odvROoENg3VoOlIh2tiEs",
  "kid": "aPctw_odvROoENg3VoOlIh2tiEs"
}.{
  "aud": "https://graph.windows.net",
  "iss": "https://sts.windows.net/2334b803-b65c-41ab-9df6-e0e2ddd82905/",
  "iat": 1573035887,
  "nbf": 1573035887,
  "exp": 1573039787,
  "acr": "1",
  "aio": "42VgYGiWSp+cKHzzUbjvqZ7fb1Wbmt8sMI+0vbauZGdbwneOFb0A",
  "amr": [
    "pwd"
  ],
  "appid": "2108a261-0ff0-490c-801f-3fd3ee0e5869",
  "appidacr": "1",
  "family_name": "",
  "given_name": "Lloyd",
  "ipaddr": "0.0.0.0",
  "name": "Lloyd",
  "oid": "121274df-131d-48e0-8eff-cc39df91f4a1",
  "onprem_sid": "S-1-5-21-3134778076-18097780-690176261-2101",
  "puid": "10032000438936F6",
  "scp": "Calendars.Read offline_access openid profile User.Read User.ReadBasic.All",
  "sub": "QCmLDfZVQdWTziFgawHg76E5wjI5cgLY-Q9v7QkrMUY",
  "tenant_region_scope": "EU",
  "tid": "2334b803-b65c-41ab-9df6-e0e2ddd82905",
  "unique_name": "Lloyd@XXXXXX.tech",
  "upn": "Lloyd@XXXXXX.tech",
  "uti": "UsrCN9Q7Bk60gPMKL8WIAA",
  "ver": "1.0"
}.[Signature]

```

**Comment by OPVL** [comment]
To get everything up and running for now, this is what I've implemented however being able to do some Graph API stuff would definitely be preferable.
```php
    protected function success(\Illuminate\Http\Request $request, $access_token, $refresh_token, $profile)
    {
        // Kinda bastardised the method however this will do for now.
        $email = strtolower($profile->upn);

        $user = User::updateOrCreate(['email' => $email], [
            'first_name' => $profile->given_name,
            'last_name' => $profile->family_name,
            'email_verified_at' => Carbon::now(),
            'azure_id' => $profile->oid // the Immutable Object Identifier eg. the user id for MS apps. (reverse lookups n ting)
        ]);

        Auth::login($user, true);

        return parent::success($request, $access_token, $refresh_token, $profile);
    }
```

**Comment by danieltjewett** [comment]
Hi @OPVL 

Sorry for the delay -- I was offline yesterday.

We haven't tested `common` before -- I'm glad that works.  Perhaps the `access_token` that comes from a `common` request is unauthorized?

I want to check two things:

1.  Just to confirm, this example from extended installation did not work?

```
<?php

namespace App\Http\Middleware;

use RootInc\LaravelAzureMiddleware\Azure as Azure;
use Microsoft\Graph\Graph;
use Microsoft\Graph\Model;

use Auth;

use App\User;

class AppAzure extends Azure
{
    protected function success($request, $access_token, $refresh_token, $profile)
    {
        $graph = new Graph();
        $graph->setAccessToken($access_token);
        
        $graph_user = $graph->createRequest("GET", "/me")
                      ->setReturnType(Model\User::class)
                      ->execute();
        
        $email = strtolower($graph_user->getUserPrincipalName());

        $user = User::updateOrCreate(['email' => $email], [
            'firstName' => $graph_user->getGivenName(),
            'lastName' => $graph_user->getSurname(),
        ]);

        Auth::login($user, true);

        return parent::success($request, $access_token, $refresh_token, $profile);
    }
}
```

2. Do we have `AZURE_SCOPE` set?  This could simply be `user.read`

**Comment by OPVL** [comment]
Hi no worries I've been pulled from pillar to post over the last few days so I've been on and off projects quite quickly. 

So my original issue was due to resource access following authentication, the permissions I'm requesting are as follows 
``` 
AZURE_RESOURCE=https://graph.windows.net
AZURE_SCOPE='User.ReadBasic.All openid User.Read' 
```
I was using the above snippet, I have reverted back to it this morning to continue testing and was able to extract my access token and perform a standard call in postman to get the un-truncated message.
```json
{
  "error": {
    "code": "InvalidAuthenticationToken",
    "message": "Access token validation failure. Invalid audience.",
    "innerError": {
      "request-id": "142d2675-6051-4350-9e9a-77780ee11876",
      "date": "2019-11-11T10:25:17"
    }
  }
}
```
I'm definitely able to sign in successfully as I'm getting a fleshed out user profile returned which I'm currently using to create my user records. It's when I'm performing a follow-up call against the graph API using the access token I got from signing in that I run in to this above error. I thought it could be due to an issue with .net|.com as the graph package uses .com and I'm requesting .net however when I try and set my resource to .com it fails in the first authentication step.

**Comment by OPVL** [comment]
Hi, just to update my previous comment.
I've now taken back the grant code that is passed back to the application after successful signing in. I've used Postman to create a basic request to the following route.

POST: https://login.microsoftonline.com/common/oauth2/token
```
grant_type:authorization_code
client_id:2108a261-ID
client_secret:SECRET
code:GRANT_CODE
resource:https://graph.windows.com
```
Which returned the following response:
```json
{
    "error": "invalid_resource",
    "error_description": "AADSTS500011: The resource principal named https://graph.windows.com was not found in the tenant named TENANT_ID. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.\r\nTrace ID: 3c05a8ec-3743-488f-9320-645d21547700\r\nCorrelation ID: 7bc06b33-fbd2-4c73-a845-443c117e1394\r\nTimestamp: 2019-11-11 10:51:32Z",
    "error_codes": [
        500011
    ],
    "timestamp": "2019-11-11 10:51:32Z",
    "trace_id": "3c05a8ec-3743-488f-9320-645d21547700",
    "correlation_id": "7bc06b33-fbd2-4c73-a845-443c117e1394",
    "error_uri": "https://login.microsoftonline.com/error?code=500011"
}
```

I believe this is directly related to the use of common here instead of my tenant ID.

**Update as I'm Writing this**
Okay so I've gotten it to work now (I think) and the workaround is to use the .com/common to sign in with and then for the second stage where I get my access token & call the graph API I need to go to .com/tenant-id/oauth2/token with my grant from the first stage of authentication.

If I can reliably get the tenant ID (need to check what profile returns) then I'm going to put this in as a config item for multi-tenanted applications and create PR so that it if true it will use the common route for the sign-in and then the tenant-id for the second stage.

Leaving open just until I verify all of this functionality is working as expected.

**Comment by OPVL** [comment]
Okay final follow up. Turns out I'd been making it much harder than I needed to. 
When using with /common endpoint, don't specify a resource and it will handle it for you.

In summary, the only difference between using as a single tenant application and multi tenant application is providing `common` instead of a tenant id in the .env

Closing as I've found a solution. Will make a PR and add this to readme as an unsupported use case.
